src/backend/distributed/commands/call.c

```diff
@@ -0,0 +1,127 @@
+/*-------------------------------------------------------------------------
+ *
+ * call.c
+ *    Commands for call remote stored procedures.
+ *
+ * Copyright (c) 2019, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+
+#include "postgres.h"
+
+#include "distributed/colocation_utils.h"
+#include "distributed/commands.h"
+#include "distributed/connection_management.h"
+#include "distributed/master_metadata_utility.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/remote_commands.h"
+#include "distributed/shard_pruning.h"
+#include "distributed/worker_manager.h"
+#include "nodes/parsenodes.h"
+#include "nodes/primnodes.h"
+#include "tcop/dest.h"
+#include "utils/syscache.h"
+
+
+static bool CallFuncExprRemotely(DistObjectCacheEntry *procedure,
+								 FuncExpr *funcExpr, const char *queryString,
+								 DestReceiver *dest);
+
+
+/*
+ * CallDistributedProcedure calls a stored procedure on the worker if possible.
+ */
+bool
+CallDistributedProcedureRemotely(CallStmt *callStmt, const char *queryString,
+								 DestReceiver *dest)
+{
+	DistObjectCacheEntry *procedure = NULL;
+	FuncExpr *funcExpr = callStmt->funcexpr;
+	Oid functionId = funcExpr->funcid;
+
+	procedure = LookupDistObjectCacheEntry(PROCOID, functionId, 0);
+	if (procedure == NULL)
+	{
+		return false;
+	}
+
+	return CallFuncExprRemotely(procedure, funcExpr, queryString, dest);
+}
+
+
+/*
+ * CallFuncExprRemotely calls a procedure of function on the worker if possible.
+ */
+static bool
+CallFuncExprRemotely(DistObjectCacheEntry *procedure, FuncExpr *funcExpr,
+					 const char *queryString, DestReceiver *dest)
+{
+	Oid colocatedRelationId = InvalidOid;
+	Node *partitionValue = NULL;
+	List *shardList = NIL;
+	ShardInterval *shard = NULL;
+	List *placementList = NIL;
+	ListCell *placementCell = NULL;
+	WorkerNode *preferredWorkerNode = NULL;
+	int connectionFlags = 0;
+	MultiConnection *connection = NULL;
+
+	if (IsMultiStatementTransaction())
+	{
+		ereport(DEBUG2, (errmsg("cannot push down CALL in multi-statement transaction")));
+		return false;
+	}
+
```

Comment:
some other things to maybe watch out for (not sure if these are actual issues) 
- Expressions in the parameters that cannot be evaluated on the worker 
- Implicit casts in the distribution argument

Reply:
The logic now only allows for cases when all arguments are `Const` nodes